src/replication.c

```diff
@@ -1914,6 +1943,169 @@ disconnect_nonresponding_replica(replica_t *replica, uint64_t io_seq,
 	MTX_UNLOCK(&replica->r_mtx);
 }
 
+static void
+handle_snap_list_response(replica_t *replica, void *resp_data, mgmt_cmd_t *mgmt_cmd, uint64_t len)
+{
+	zvol_io_hdr_t *hdr = replica->mgmt_io_resp_hdr;
+	struct zvol_snapshot_list *snap_details;
+	rcommon_mgmt_cmd_t *rcomm_mgmt = mgmt_cmd->rcomm_mgmt;
+	bool delete = false, assigned, failed = false;
+	struct zvol_snapshot_list **snap_resp_array = (struct zvol_snapshot_list **)rcomm_mgmt->buf;
+	uint64_t i = 0;
+
+	snap_details = calloc(1, len);
+	memcpy(snap_details, resp_data, len);
+	free(resp_data);
+
+	MTX_LOCK(&rcomm_mgmt->mtx);
+	if (hdr->status != ZVOL_OP_STATUS_OK || snap_details->error) {
+		rcomm_mgmt->cmds_failed++;
+		free(snap_details);
+		failed = true;
+	} else {
+		rcomm_mgmt->cmds_succeeded++;
+	}
+
+	if ((rcomm_mgmt->caller_gone == 1) &&
+	    (rcomm_mgmt->cmds_sent == (rcomm_mgmt->cmds_failed + rcomm_mgmt->cmds_succeeded)))
+		delete = true;
+
+	if (failed)
+		goto out;
+
+	for (i = 0, assigned = false;
+	    i < (rcomm_mgmt->buf_size / sizeof (*snap_resp_array));
+		i++) {
+		if (snap_resp_array[i] == NULL) {
+			snap_resp_array[i] = (struct zvol_snapshot_list *)snap_details;
+			assigned = true;
+			break;
+		}
+	}
+
+	ASSERT(assigned);
+
+out:
+
+	MTX_UNLOCK(&rcomm_mgmt->mtx);
+
+	if (delete == true) {
+		DESTROY_SNAPSHOT_RESP_LIST(rcomm_mgmt);
+		free_rcommon_mgmt_cmd(rcomm_mgmt);
+	}
+}
+
+static char *
+get_replica_id_for_guid(spec_t *spec, uint64_t guid)
+{
+	replica_t *replica;
+	char *replica_id = NULL;
+
```

Comment:
having an assert on spec->rq_mtx would be good

Reply:
done